Update docs that Parameters are immune to no_grad mode

janeyx99 opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/83998

![image](https://user-images.githubusercontent.com/31798555/220971800-4af57d92-9f15-4e13-bfe4-73e2ff1cd943.png)
![image](https://user-images.githubusercontent.com/31798555/221019508-d7330a16-7f01-4d37-a1af-a4905e9596c4.png)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95232

janeyx99 committed:
Update docs that Parameters are immune to no_grad mode

[ghstack-poisoned]

janeyx99 committed:
Update on "Update docs that Parameters are immune to no_grad mode"




[ghstack-poisoned]

janeyx99 committed:
Update on "Update docs that Parameters are immune to no_grad mode"

Fixes https://github.com/pytorch/pytorch/issues/83998




[ghstack-poisoned]

soulitzer commented:
Why do we need to explicitly point this out? Tensors created with requires_grad=True will be unaffected by no-grad whether or not they are Parameters.
```
>>> with torch.no_grad():
...   a = torch.tensor(1., requires_grad=True)
...
>>> a
tensor(1., requires_grad=True)
```

janeyx99 commented:
How do you think we should best phrase to make sure users would not be surprised like in https://github.com/pytorch/pytorch/issues/83998 then?

soulitzer commented:
> Should I mention that it is broader than just Parameters?

Sounds good to me!

janeyx99 committed:
Update on "Update docs that Parameters are immune to no_grad mode"

Fixes https://github.com/pytorch/pytorch/issues/83998

![image](https://user-images.githubusercontent.com/31798555/220462533-c2b6d1d1-a811-4897-b889-354aeb0d7557.png)
![image](https://user-images.githubusercontent.com/31798555/220462646-164c29d5-ce0f-4ecd-96bb-649d2e80525c.png)




[ghstack-poisoned]

soulitzer commented on torch/autograd/grad_mode.py:
```diff
@@ -14,7 +14,11 @@ class no_grad(_DecoratorContextManager):
     In this mode, the result of every computation will have
-    `requires_grad=False`, even when the inputs have `requires_grad=True`.
+    `requires_grad=False`, even when the inputs have `requires_grad=True`,
+    UNLESS the result is a tensor created with `requires_grad=True`,
```
nit: it might not be clear that this is applies to new tensor construction, since we're talking about doing "computation". Maybe we could explicitly point out the tensor construction?

albanD replied:
nit: I think this can be a second sentence. Also mentioning that the exception are 'factory functions' is good: all functions that create a new Tensor and take a requires_grad kwarg.


Also you `requires_grad=True` kwarg is exactly the same as the attribute `requires_grad=True` above. Given that they refer to very different things they shouldn't be the same.

soulitzer commented on torch/nn/parameter.py:
```diff
@@ -25,7 +25,10 @@ class Parameter(torch.Tensor, metaclass=_ParameterMeta):
         data (Tensor): parameter tensor.
         requires_grad (bool, optional): if the parameter requires gradient. See
-            :ref:`locally-disable-grad-doc` for more details. Default: `True`
+            :ref:`locally-disable-grad-doc` for more details. Note that the
```
nit: put the link after the "Note that..." sentence

janeyx99 committed:
Update on "Update docs that Parameters are immune to no_grad mode"

Fixes https://github.com/pytorch/pytorch/issues/83998

![image](https://user-images.githubusercontent.com/31798555/220462533-c2b6d1d1-a811-4897-b889-354aeb0d7557.png)
![image](https://user-images.githubusercontent.com/31798555/220462646-164c29d5-ce0f-4ecd-96bb-649d2e80525c.png)




[ghstack-poisoned]

soulitzer reviewed (approved):
Thanks!

janeyx99 commented:
@pytorchbot merge -f "docs change only, slow test failure irrelevant"

ZainRizvi commented:
@pytorchbot revert -m "This caused the test_doc_examples test to fail on trunk" -c nosignal

Hey @janeyx99, turns out we run mypy on trunk to validate doc changes, and this PR caused that validation to fail ([logs](https://hud.pytorch.org/pytorch/pytorch/commit/5783cee2a3a1457fc93b00a4a50e61ba02f148db)).

cc @clee2000 to see if we can move that check to the pull workflow.

clee2000 commented:
> @pytorchbot revert -m "This caused the test_doc_examples test to fail on trunk" -c nosignal
> 
> Hey @janeyx99, turns out we run mypy on trunk to validate doc changes, and this PR caused that validation to fail ([logs](https://hud.pytorch.org/pytorch/pytorch/commit/5783cee2a3a1457fc93b00a4a50e61ba02f148db)).
> 
> cc @clee2000 to see if we can move that check to the pull workflow.

this test also exists in pull and failed

janeyx99 commented:
yea sorry this was a hasty disregard on my part!

janeyx99 committed:
Update on "Update docs that Parameters are immune to no_grad mode"

Fixes https://github.com/pytorch/pytorch/issues/83998

![image](https://user-images.githubusercontent.com/31798555/220971800-4af57d92-9f15-4e13-bfe4-73e2ff1cd943.png)
![image](https://user-images.githubusercontent.com/31798555/220971892-35554d17-fc44-4211-9017-7a5555ae3bb1.png)




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
